### PR TITLE
Restrict named pipes transport to be Windows only

### LINKED ISSUE #46231: Disable named pipes transport on non-Windows platforms
Kestrel now supports a named pipes transport. However, the .NET named pipes API supports Windows and Unix/macOS. On those platforms, the win32 named pipes API is replaced with Unix domain sockets. There are subtle behavior differences.

For simplicity, I think we should disable named pipes transport on non-Windows platforms. Someone who wants to be cross-platform can use UDS directly with `ListenUnixSocket`.

* Don't add named pipe transport or service if not on Windows. Either throw an error or write a log message. Probably throw an error.
* Place the OS-supported attribute with Windows on `KestrelServerOptions.ListenNamedPipe` API?
* Skip all existing named pipe transport tests
* Add test that verifies that adding a named pipe transport fails on non-Windows platforms.

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/46231

### CODE REVIEW

**Review by mgravell — APPROVED** [review]

**Comment by mgravell** [comment]
lgtm; I was searching around for "what's that attribute you can add ..." - found `[SupportedOSPlatform]` then (and only then) noticed you already had it; d'oh!

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs [hunk 1]
```diff
@@ -53,6 +53,14 @@ public async Task<EndPoint> BindAsync(EndPoint endPoint, ConnectionDelegate conn
             }
         }
 
+        // Special case situation where a named pipe endpoint is specified and there is no matching transport.
+        // The named pipe transport is only registered on Windows. The test is done at this point so there is
+        // the opportunity for the app to register their own transport to handle named pipe endpoints.
```
Comment (BrennanConroy) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs:58]: Should there be a test that registers its own "named pipe" transport to make sure we don't throw?

File: src/Servers/Kestrel/Kestrel/src/WebHostBuilderKestrelExtensions.cs [hunk 2]
```diff
@@ -46,7 +46,10 @@ public static IWebHostBuilder UseKestrel(this IWebHostBuilder hostBuilder)
             options.DefaultCloseErrorCode = (long)Http3ErrorCode.NoError;
         });
 
-        hostBuilder.UseNamedPipes();
+        if (OperatingSystem.IsWindows())
+        {
+            hostBuilder.UseNamedPipes();
```
Comment (danmoseley) [inline src/Servers/Kestrel/Kestrel/src/WebHostBuilderKestrelExtensions.cs:51]: this applies to the original, but for my education, does this line mean named pipes are simply enabled as an option, not that they are somehow the default transport?
(if the latter, perhaps EnableNamedPipes() would be a better name, but perhaps UseXX() is the established pattern)

Comment (JamesNK) [reply src/Servers/Kestrel/Kestrel/src/WebHostBuilderKestrelExtensions.cs:51]: UseXXX is the established pattern. AddXXX would be better but the current pattern dates from when there was one transport at a time. Stuck with UseXXX to be consistent.

**Pull request merged**